primitives/src/v3/mod.rs

```diff
@@ -0,0 +1,1804 @@
+		input: &mut I,
+	) -> Result<Self, parity_scale_codec::Error> {
+		let maybe_magic = <[u8; 4]>::decode(input)?;
+		if maybe_magic != BACKING_STATEMENT_MAGIC {
+			return Err(parity_scale_codec::Error::from("invalid magic string"))
+		}
+
+		Ok(match CompactStatementInner::decode(input)? {
+			CompactStatementInner::Seconded(h) => CompactStatement::Seconded(h),
+			CompactStatementInner::Valid(h) => CompactStatement::Valid(h),
+		})
+	}
+}
+
+impl CompactStatement {
+	/// Get the underlying candidate hash this references.
+	pub fn candidate_hash(&self) -> &CandidateHash {
+		match *self {
+			CompactStatement::Seconded(ref h) | CompactStatement::Valid(ref h) => h,
+		}
+	}
+}
+
+/// `IndexedVec` struct indexed by type specific indices.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub struct IndexedVec<K, V>(Vec<V>, PhantomData<fn(K) -> K>);
+
+impl<K, V> Default for IndexedVec<K, V> {
+	fn default() -> Self {
+		Self(vec![], PhantomData)
+	}
+}
+
+impl<K, V> From<Vec<V>> for IndexedVec<K, V> {
+	fn from(validators: Vec<V>) -> Self {
+		Self(validators, PhantomData)
+	}
+}
+
+impl<K, V> FromIterator<V> for IndexedVec<K, V> {
+	fn from_iter<T: IntoIterator<Item = V>>(iter: T) -> Self {
+		Self(Vec::from_iter(iter), PhantomData)
+	}
+}
+
+impl<K, V> IndexedVec<K, V>
+where
+	V: Clone,
+{
+	/// Returns a reference to an element indexed using `K`.
+	pub fn get(&self, index: K) -> Option<&V>
+	where
+		K: TypeIndex,
+	{
+		self.0.get(index.type_index())
+	}
+
+	/// Returns number of elements in vector.
+	pub fn len(&self) -> usize {
+		self.0.len()
+	}
+
+	/// Returns contained vector.
+	pub fn to_vec(&self) -> Vec<V> {
+		self.0.clone()
+	}
+
+	/// Returns an iterator over the underlying vector.
+	pub fn iter(&self) -> Iter<'_, V> {
+		self.0.iter()
+	}
+
+	/// Returns a mutable iterator over the underlying vector.
+	pub fn iter_mut(&mut self) -> IterMut<'_, V> {
+		self.0.iter_mut()
+	}
+
+	/// Creates a consuming iterator.
+	pub fn into_iter(self) -> IntoIter<V> {
+		self.0.into_iter()
+	}
+
+	/// Returns true if the underlying container is empty.
+	pub fn is_empty(&self) -> bool {
+		self.0.is_empty()
+	}
+}
+
+/// The maximum number of validators `f` which may safely be faulty.
+///
+/// The total number of validators is `n = 3f + e` where `e in { 1, 2, 3 }`.
+pub fn byzantine_threshold(n: usize) -> usize {
+	n.saturating_sub(1) / 3
+}
+
+/// The supermajority threshold of validators which represents a subset
+/// guaranteed to have at least f+1 honest validators.
+pub fn supermajority_threshold(n: usize) -> usize {
+	n - byzantine_threshold(n)
+}
+
+/// Information about validator sets of a session.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(PartialEq))]
+pub struct SessionInfo {
+	/****** New in v3 *******/
+	/// Executor parameter set for the session.
+	pub executor_params: ExecutorParams,
```

Comment:
Can we please move this to the bottom? Then we don't require the entire ecosystem to upgrade before the runtime upgrade hits on chain. However, this needs to be tested!

Reply:
What a beautiful and elegant solution, and moreover, it works, I've tested it!
However, after a discussion, we decided to proceed with a more radical one: not to include `ExecutorParams` into `SessionInfo` at all. We'll be holding it aside. Too much may be broken due to this change, and it's better to be on the safe side.
See also #6586

Reply:
Why does that work? 🤔

> However, after a discussion, we decided to proceed with a more radical one: not to include ExecutorParams into SessionInfo at all. We'll be holding it aside. Too much may be broken due to this change, and it's better to be on the safe side.
See also https://github.com/paritytech/polkadot/issues/6586

To be clear, that means you are doing this (from #6586)? 

> We can also be more relaxed and leave SessionInfo alone, but stop modifying it. Anything new we need, any modifications we need should be their own new runtime API call.

Reply:
> Why does that work? :thinking: 

Say you have `struct A { x: u8, y: u8 }` and `struct B { x: u8 }`. Then `A { x: 1, y: 2 }` would be encoded by SCALE codec into something like `[ 'x', 1, 'y', 2 ]` (shitty pseudocode, I know). If we try to decode that stream into `B`, we'll still get a valid `B { x: 1 }`, and the tail of the stream is just ignored. Neither did I know it works like that.

> To be clear, that means you are doing this (from #6586)?

I'm rolling back to the state when we had `ExecutorParams` in a separate storage map, not encapsulated into `SessionInfo`, and a separate runtime call to retrieve it. Then I'm re-applying all the useful changes made after integration with `SessionInfo`. I'm not demolishing `SessionInfo` itself, not in the scope of this PR at least.